### DISCUSSION TITLE
Remove lines and noise in the image

### DISCUSSION THREAD

**Discussion by jueqing1015** [dlemstra/Magick.NET#667] [Help] [answered]
I want to use MAGICK.NET to delete the lines and noises in the image
Try Morphology but the effect is not too obvious

**Comment by dlemstra** [comment]
The links to your images appear to be broken.

**Reply by jueqing1015** [reply]
I don’t know why the image link I uploaded always fails, I created a compressed package, you download it and take a look
[sample-0005 - 副本 (2).zip](https://github.com/dlemstra/Magick.NET/files/4778281/sample-0005.-.2.zip)

**Comment by dlemstra** [comment]
I think we need the expert @fmw42 to help us with this. If he can tell us how to do it on the command line I can help you translate it to c#.

**Reply by jueqing1015** [reply]
@dlemstra 
Hello, can you help me translate to C#?

**Comment by fmw42** [comment]
The best way would be to use connected components to remove very small and very large regions.

This leaves only a couple of small black spots:
```
convert -quiet sample-0005-original.jpg \
-threshold 0 -type bilevel \
-define connected-components:exclude-header=true \
-define connected-components:area-threshold=3-1000 \
-define connected-components:mean-color=true \
-connected-components 8 \
sample-0005_result1.png

```

![sample-0005_result1](https://user-images.githubusercontent.com/13458218/84689266-95b35580-aef5-11ea-9506-ded595914a5a.png)


The removes those spots, but fills in white holes in some letters such as "e"
```
convert -quiet sample-0005-original.jpg \
-threshold 0 -type bilevel \
-define connected-components:exclude-header=true \
-define connected-components:area-threshold=4-1000 \
-define connected-components:mean-color=true \
-connected-components 8 \
sample-0005_result2.png

```

![sample-0005_result2](https://user-images.githubusercontent.com/13458218/84689275-9a780980-aef5-11ea-9e5b-8791c12253f0.png)

**Comment by dlemstra** [comment]
You currently cannot set a range for the area threshold but you will be able to do this when version `7.20.0.0` has been released:

```C#
using (var image = new MagickImage(@"sample-0005 - 副本 (2).jpg"))
{
    image.Threshold(new Percentage(0)); // -threshold 0
    image.ColorType = ColorType.Bilevel; // -type bilevel

    var settings = new ConnectedComponentsSettings()
    {
        AreaThreshold = new Threshold(3.05, 1000), // -define connected-components:area-threshold=3.05-1000
        MeanColor = true, // -define connected-components:mean-color=true
        Connectivity = 8, // -connected-components 8
    };
    image.ConnectedComponents(settings); // -connected-components 8

    image.Write("sample-0005-result.jpg");
}
```

I used `3.05` instead of `4` and got a slightly better result for this image.

**Reply by jueqing1015** [reply]
Ok i will try after the new version is released

**Reply by jueqing1015** [reply]
@dlemstra 

Hello, I download the new version to test
But running to this sentence
image.ConnectedComponents(settings);
Will report an error
The error is as follows
too many objects `c:\84689266-95b35580-aef5-11ea-9506-ded595914a5a.png' @ error/vision.c/ConnectedComponentsImage/437
The image address is:
https://user-images.githubusercontent.com/13458218/84689266-95b35580-aef5-11ea-9506-ded595914a5a.png

**Reply by dlemstra** [reply]
Are you using the Q8 or Q16 version of Magick.NET? I suspect that you are getting to many areas. Maybe @fmw42 knows a trick to make this work in Q8?

**Reply by jueqing1015** [reply]
I use Magick.NET-Q8-AnyCPU
The strange thing is that some images do not report errors and some report errors.

**Reply by fmw42** [reply]
Q8 will only allow 255 individual regions in your thresholded image. So you may have to use morphology to combine regions before using connected components. Q16 would allow 65535 regions. So you may want to consider upgrading to Q16 ImageMagick install.

**Reply by jueqing1015** [reply]
@dlemstra 
If upgrading to the Q16 version, will it cause the entire process to be slow, because I see the difference between the two. Q16 requires a larger memory space

**Reply by fmw42** [reply]
Yes, it will be a bit slower due to the larger memory usage.

**Reply by jueqing1015** [reply]
It may not be able to upgrade, because the overall speed of Q16 will slow down the entire process. This will affect the efficiency a bit. Thank you!

**Comment by fmw42** [comment, marked as answer]
Text images like the one above have lots of characters and thus lots of isolated regions. If there are more than 255 characters (including dots on i), then connected components will limit out for Q8. That is the trade you will have to assess vs using Q16. Your processing will likely also slow down if you are using HDRI compile.
